Predicate cacheing and cleanup

jayunit100 opened the pull request:
Fix to #31795 

First pass @ cleanup and caching of the CheckServiceAffinity function.  

The cleanup IMO is necessary because the logic around the pod listing and the use of the "implicit selector" (which is reverse engineered to enable the homogenous pod groups).

Should still pass the E2Es.

@timothysc @wojtek-t

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33763)

<!-- Reviewable:end -->

jayunit100 commented:
@k8s-bot gke e2e test this #33175

jayunit100 commented:
@k8s-bot gke e2e test this.

jayunit100 commented:
@k8s-bot gke e2e test this.

spxtr commented:
@k8s-bot gke test this https://github.com/kubernetes/test-infra/issues/709

jayunit100 commented:
still not sure why e2es are dying here

timothysc commented:
So instead of breaking it up into functional objects, why not create a series of well defined utility functions?  I see most of those f(n)'s as being generally useful and it would also be easier to read.

jayunit100 commented:
Broader context for Utils, data access patterns and so on.  Bee have something like a dao wrapper query lib in e2e/federated/... maybe time to bump it to core if the same thing doesn't already exist there?

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -73,6 +73,9 @@ type predicateMetadata struct {
 	matchingAntiAffinityTerms []matchingPodAntiAffinityTerm
+	// Service pods cache, otherwise this is a large query, will slow things down
+	// when there are 100s of services/pods.
+	servicePods []*api.Pod
```
This is currently not set anywhere.

jayunit100 replied:
predicateMeta.servicePods caster out of meta,  are you saying that the modification to the pointer won't persist?

jayunit100 replied:
will dig into this and see , maybe I can update the unit tests to uncover that race

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -637,82 +640,115 @@ func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister al
+// 		- L is not defined in the pod.
+// 		- AND SOME pod from a service is the same namespace, already scheduled onto a node provides that value.
 func (s *ServiceAffinity) CheckServiceAffinity(pod *api.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+	predicateMeta := meta.(*predicateMetadata)
```
Please use the pattern like this to make it easily usable in unit tests:
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L458

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -637,82 +640,115 @@ func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister al
+			}
+		}
+		return filtered
+	}
```
I definitely agree with @timothysc  - those function aren't special for this predicate. They should be at least living alone functions in this file.

jayunit100 replied:
Do we have a home for queries and data access functions anywhere else...? If so maybe I can just reuse entirely... if not maybe we could start one....

jayunit100 commented:
heres another impl.  Shud be passing assume failures are flakes

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -36,6 +36,10 @@ import (
 )
 
+const (
+	PREDICATE_METADATA_REFRESH = time.Minute
```
We use camelcase convention for everything.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -73,6 +77,12 @@ type predicateMetadata struct {
+	// when there are 100s of services/pods.
+	lock                       sync.Mutex
+	servicePods                []*api.Pod
+	servicePodsLastRefreshTime time.Time
```
I don't understand the logic behind refreshing it. Basically, metadata should be precoumputed for a given pod before starting computing predicates for each nodes and we should use exactly the same data for each nodes.

jayunit100 replied:
Was assuming at any given point this data could get stale but probably that is ok given your comment

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -637,82 +647,79 @@ func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister al
+// 		- AND SOME pod from a service is the same namespace, already scheduled onto a node provides that value.
 func (s *ServiceAffinity) CheckServiceAffinity(pod *api.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+	if _, ok := meta.(*predicateMetadata); !ok {
+		panic("Failed casting predicate metadata")
```
Please don't use panics in the code.

wojtek-t replied:
Can you really please use the construct from here:
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L458

?

jayunit100 replied:
sure; if i just save a local copy of the cached list . i can impl that easily , will update shortly

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/utils.go:
```diff
@@ -0,0 +1,52 @@
+package predicates
```
Boilerplate is missing.

jayunit100 commented:
ok i added unit tests

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/utils_test.go:
```diff
@@ -0,0 +1,68 @@
+
+	fmt.Println(labelSubset)
+
+	fmt.Println(filterPodsByNamespace(nsPods, "ns1"))
```
This doesn't seem like a test - can you please ping me when this is ready for review.

jayunit100 replied:
hows that ? We use something similar in the controller packages ... https://blog.golang.org/examples , should run in the CI and verify the comments.

wojtek-t replied:
That is super interesting - I completely wasn't aware of it. Sorry for this comment then (and thanks for the link pointer).

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -73,6 +73,11 @@ type predicateMetadata struct {
+
+	// CheckServiceAffinity: pods cache, otherwise this is a large query, will slow things down
+	// when there are 100s of services/pods.
+	lock        sync.Mutex
```
I'm really not a fan of this approach. I think that really instead of it, we should pass something like names of all predicates to "PredicateMetada" function:
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/generic_scheduler.go#L162

and based on the fact if the ServiceAffinity is registered or not precompute servicePods.

jayunit100 replied:
hmmm, i guess if we make it a generic PreCompute maybe other predicates could reuse, ill think about this and see whats possible.  i was assuming lazy+decoupled would be the simplest way forward.  

My original thinking was that we may want to mutate the servicePods to nil at some point to force recompute, i.e. because there might be more than one way to compute affinity, and we might want to redo the computation in a cluster thats running.... but I guess `findNodesThatFit` will be recalled anytime pods die off, so the affinity will be naturally recalculated by the scheduler on an as needed basis.

wojtek-t replied:
Yes - findNodesThatFit will be called from scratch any time we will be scheduling a pod (including situations when e.g. pod failed to schedule and we are retrying). So the metadata will never be reused. So it's save to just precompute it.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -73,14 +73,37 @@ type predicateMetadata struct {
+
+	// CheckServiceAffinity: pods cache, otherwise this is a large query, will slow things down
+	// when there are 100s of services/pods.
+	lock        sync.Mutex
```
This lock is no longer needed now.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -73,14 +73,37 @@ type predicateMetadata struct {
-func PredicateMetadata(pod *api.Pod, nodeInfoMap map[string]*schedulercache.NodeInfo) interface{} {
+// A set of functions which take in the predicateMetadata type, and add information to it.
+// This map should only be modified when registering predicates.
+var predicatePrecomputations map[string]func(pm *predicateMetadata) = make(map[string]func(pm *predicateMetadata))
```
Please define a type for "func (*predicateMetadata)"

wojtek-t replied:
Also - this should be defined in plugin/pkg/scheduler/algorithm/types.go.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -73,14 +73,37 @@ type predicateMetadata struct {
+// This map should only be modified when registering predicates.
+var predicatePrecomputations map[string]func(pm *predicateMetadata) = make(map[string]func(pm *predicateMetadata))
+
+func RegisterPredicatePrecomputation(predicateName string, precomp func(pm *predicateMetadata)) {
```
Please add a comment that this function is not thread-safe (or make it thread-safe?)

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -89,12 +112,25 @@ func PredicateMetadata(pod *api.Pod, nodeInfoMap map[string]*schedulercache.Node
 		podRequest:                getResourceRequest(pod),
 		podPorts:                  getUsedPorts(pod),
 		matchingAntiAffinityTerms: matchingTerms,
+		servicePods:               make(map[string]([]*api.Pod)),
```
We should call this make only in precomputation functions - othewise if this is unused, we are loosing unnecessary memory allocation.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -73,14 +73,37 @@ type predicateMetadata struct {
+
+// PredicateMetadata generates whatever metadata will be used over time to process predicate logic.
+// Optional Vararg: PredicateMetadata is given access to all predicate information, so that it precompute data as necessary as an optimization.
+func PredicateMetadata(pod *api.Pod, nodeInfoMap map[string]*schedulercache.NodeInfo, predicatesOptionalArg ...map[string]algorithm.FitPredicate) interface{} {
```
I don't understand why this is variadic argument. This should just be:
```
registeredPredicates map[string]algorithm.FitPredicate
```

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -89,12 +112,25 @@ func PredicateMetadata(pod *api.Pod, nodeInfoMap map[string]*schedulercache.Node
+	if predicates != nil {
+		for predicateName, precomputationFunc := range predicatePrecomputations {
+			if predicates[predicateName] != nil {
+				glog.V(1).Info("Registering precomputation for %v", predicateName)
```
glog.V(4) - this log would be to spammy

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -73,14 +73,37 @@ type predicateMetadata struct {
+	// CheckServiceAffinity: pods cache, otherwise this is a large query, will slow things down
+	// when there are 100s of services/pods.
+	lock        sync.Mutex
+	servicePods map[string][]*api.Pod
```
Why this is map[string] ?
This should simply be:
```
servicePods []*api.Pod
```

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,92 +663,96 @@ type ServiceAffinity struct {
+	return affinity.CheckServiceAffinity, func(pm *predicateMetadata) {
+		var err error
+		// We can't know at the start what labels we are selecting against.
+		pm.servicePods[""], err = podLister.List(createSelectorFromLabels(nil))
```
This isn't what I had in mind.
Basically, we should extend predicateMetadata to contain the pod (or just its selector) and use this selector here.

As for podLister, we should use the mechanism that is already present for prioityMetadata (we should just duplicate it for predicates):
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go#L64

Basically, we should add a similar method for predicatemetadata:
"RegisterPredicateMetadataProducerFactory"
and then using the args from here we would be able to make the "podLister" be a filed of "PredicateMetadataFactory".

jayunit100 replied:
- Ah yes, since PredicateMetadata is done once per scheduling event i was wondering why it didnt retain the entire pod itself....  I assumed maybe that it was decoupled for a reason, after alll, its using other pod information.  If we preserve the whole pod that definetly simplifeis things.

- Regarding podListers... what do you mean `filed of PredicatreMetadataFactory` ?

wojtek-t replied:
OK - sorry for not being clear.
What I meant is that once you create "RegisterPredicateMetadataProducerFactory" (that would use exactly the same mechanism as "RegisterPriorityMetadataProducerFactory", you would be able to write somethig like this (in default.go):
```
factory.RegisterPredicateMetadataProducerFactory(
  func(args factory.PluginFactoryArgs) algorithm.Metadataproducer {
    return predicates.NewPredicateMetadataFactory(args.PodLister)
  }
}
```

and define the factory like:
```
type PredicateMetadataFactory struct {
   podLister algorithm.PodLister
}

func NewPredicateMetadataFactory(podLister algorithm.PodLister) algorithm.MetadataProducer {
  factory := &PredicateMetadataFactory{podLister}
  return factory.GetMetadata
}

func (factory *PredicateMetadataFactory) Getmetadata(pod *api.Pod) PredicateMetadata {
  // compute the metadata for the given pod
  // but you already have access to "podLister" here (which you basically need)
}
```

Does it make sense now?

wojtek-t replied:
Obviously my code doesn't compile, but it should show the idea.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,92 +663,96 @@ type ServiceAffinity struct {
-// The set of labels to be considered are provided to the struct (ServiceAffinity).
-// The pod is checked for the labels and any missing labels are then checked in the node
-// that hosts the service pods (peers) for the given pod.
+func (s *ServiceAffinity) UpdatePredicateMetaServicePods(selector labels.Selector, predicateMeta *predicateMetadata) error {
```
Once the above comment is applied, this will no longer be needed.

jayunit100 commented:
Looks like something on head is causing my runs of scheduler_perf to panic, not sure exactly what yet, so this is on hold for a day while i finish figuring it out

jayunit100 commented:
Ok , I think i see what you meant after playing around to get the right interface<->struct conversion between the meta fields.  should be ready for another look .  

- needs some cleanup @ the imports.  
- i will monitor ,  want to make sure i didnt break anything in e2es, as its a big patch
- Updated checkserviceaffinity unit tests here to actually create a predicate, since there is a direct coupling for precomputation - am considering making checkServiceAffinity a inner function (for now, i made it non-exported).

jayunit100 commented on plugin/pkg/scheduler/factory/plugins.go:
```diff
@@ -316,6 +323,17 @@ func getPriorityMetadataProducer(args PluginFactoryArgs) (algorithm.MetadataProd
+	defer schedulerFactoryMutex.Unlock()
+
+	if predicateMetadataProducer == nil {
+		fmt.Println("WARNING: EMpty metadataproducer !!!!!")
```
some debug i need to remove..

jayunit100 commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -149,7 +150,19 @@ func findNodesThatFit(
+) ([]*api.Node, FailedPredicateMap, error) {
+
+	/**
+	 * PredicateMetadata is an internal package of predicates.  We don't want or need to know the type.
```
I think i will delete this comment, since the function return is what is really important

jayunit100 commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -0,0 +1,69 @@
+
+	if predicatePrecomputations != nil {
+		for predicateName, preComputeFunction := range predicatePrecomputations {
+			glog.V(4).Info("Precompute: %v", predicateName)
```
i probably will delete this log entirely for speed.  this can be called thousands of times

jayunit100 commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -0,0 +1,69 @@
+/*
```
@wojtek-t i *could* break it up, but my first question would be - how should we separate the changes? 

- the metadata.go changes  effect predicates.go, because now predicates.go doesnt implement the CheckService algorithm
- the changes to types.go are only meaningfull in the context of enabling matching pod metadata / node information when calcualting predicates via a factory
- the changes to plugins.go are simply machinery to enable the predicates.go metadata to be plugged in via a factory
-  also the changes to generic_scheduler are really only there for enabling factoy-ization

so its a pretty cohesive change set.   

Either way is ok, just let me know... but figured i would double-check that you think its necessary...

wojtek-t replied:
I think that what you can separate is all the changes to ServiceAffinity (which means all the changes in "checkServiceAffinity" and the utils and utils_test.go files). This would significantly reduce the size of this PR.
Do you agree?

jayunit100 replied:
hmmm... all of the changes to checkServiceAffinity are dependent on precomputation working properly though :) .   without it, checkServiceAffinity will still compile - but it wont ever have a complete pod affinity cache, so the match will  return true for all nodes, because it wont see any affinity constraints, and then the unit tests would fail..

jayunit100 committed:
PredicateMetadata factory and optimization, Cleaned up some comments,
Comments addressed, Make emptyMetadataProducer a func to avoid casting,
FakeSvcLister: remove error return for len(svc)=0.  New test for
predicatePrecomp to make method semantics explictly enforced when meta
is missing. Precompute wrapper.

jayunit100 commented:
(tagged as WIP so that we dont review - im breaking this apart)

jayunit100 commented:
gated on #34624

jayunit100 commented:
Just rebased again against the gated PR.  
The second commit now is much smaller

jayunit100 commented:
Update on this: Im going to do a final review and rebase tomorrow morning then will ping for final look to save time for everyone.  A couple last minute things I want to add.

wojtek-t commented:
@jayunit100 - can you please rebase?

jayunit100 commented:
(rebased and updated, ready for review)

jayunit100 commented:
I applied these and the original changes to the master from just before scheduler_perf regressed,and I setup an A/B test comparing performance where we do vs. do not do the extra checkServiceAffinity query.  Looks like a 10x improvement.

When we rely on predicateMetadata, 
```
     0.75s 0.022% 91.83%     31.13s  0.92%  k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates.(*ServiceAffinity).checkServiceAffinity
```
When we don't rely on predicateMetadata, and keep the intermediate computation for every pod... 
```
     1.32s  0.04% 92.22%    387.43s 11.83%  k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates.(*ServiceAffinity).checkServiceAffinity
```
Conclusion = removing this from checkServiceAffinity significantly speeds the function up without breaking functionality.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -0,0 +1,70 @@
+
+	// precomputations are to be registered when predicates are constructed, these are for storing/precomputing data
+	// when a predicated is first created, to avoid having to check / lock / lazy update when predicate is being
+	// evaluated.
```
To be honest, I'm completely lost in this comment. Can you please clarify?

jayunit100 replied:
I have a better idea.  I can consolidate the precomputations info into this same file, so no comment is needed, and less coupling ..

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -0,0 +1,70 @@
+	// when a predicated is first created, to avoid having to check / lock / lazy update when predicate is being
+	// evaluated.
+	if predicatePrecomputations != nil {
+		for predicateName, preComputeFunction := range predicatePrecomputations {
```
nit: s/preCompute/precompute/ ?

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -67,34 +67,40 @@ func (c *CachedNodeInfo) GetNodeInfo(id string) (*api.Node, error) {
+/**
+*  Note that predicateMetdata and matchingPodAntiAffinityTerm need to be declared in the same file
+*  due to the way declarations are processed in predicate declaration unit tests.
+ */
```
We are generally trying to avoid such style comments. Please change to multiple "// coment" lines.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -67,34 +67,40 @@ func (c *CachedNodeInfo) GetNodeInfo(id string) (*api.Node, error) {
 }
 
+/**
+*  Note that predicateMetdata and matchingPodAntiAffinityTerm need to be declared in the same file
```
really? Isn't it enough for them to be in the same package?

jayunit100 replied:
not sure but, iirc it failed compilation when it was  in metadata.go (which is in the same predicates package) because theres some kind of parsing step that we do... is this a bug ?  i can double verify if necessary.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -67,34 +67,40 @@ func (c *CachedNodeInfo) GetNodeInfo(id string) (*api.Node, error) {
 	podPorts                  map[int]bool
 	matchingAntiAffinityTerms []matchingPodAntiAffinityTerm
-}
+	nodeNameToInfo            map[string]*schedulercache.NodeInfo
```
Do we really need it? I wouldn't like to have it here.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -67,34 +67,40 @@ func (c *CachedNodeInfo) GetNodeInfo(id string) (*api.Node, error) {
+// predicatePrecomputations is a set of functions which take in the predicateMetadata type.
+// Note that these functions mutate predicateMetadata, so it should stay private and only be used when
+// creating predicateMetadata.
+var predicatePrecomputations map[string]func(pm *predicateMetadata) = make(map[string]func(pm *predicateMetadata))
```
Please define "func(pm *predicateMetadata)" as a type

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,36 +633,60 @@ type ServiceAffinity struct {
+
+		// Precompute function: Run once by the scheduler before looping through the Predicate.  Alot of times nil, but
+		// in this case, its important to try to only List all service pods once, for performance.
+		func(pm *predicateMetadata) {
```
please factory this out to a separate function instead of defining it here

jayunit100 replied:
Ok, but before i do so .. why would we want to expose this ?  After all, the only purpose of this function is to provide the precomputation which is defined exclusively for this predicate and for no other purpose...

wojtek-t replied:
It doesn't have to be public. I'm just trying to split the code to self-contained parts. And this function is pretty self-contained.

jayunit100 replied:
ok... It is self contained, but it references some other functions / data (i.e. serviceLister, and so on) so its highly coupled... but i will separate it out if you think its necessary.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,36 +633,60 @@ type ServiceAffinity struct {
-// (i.e. it returns true IFF this pod can be added to this node, such
-// that all other pods in the same service are running on nodes w/
-// the exact same ServiceAffinity.label values).
+// The checkServiceAffinity predicate matches nodes in such a way to force that ServiceAffinity.labels are homogenous
```
Please revert to previous lengths of lines

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,36 +633,60 @@ type ServiceAffinity struct {
 // Otherwise:
-// 	Create an "implicit selector" which gaurantees pods will land on nodes with similar values
-//	for the affinity labels.
+// 	Create an "implicit selector" which gaurantees pods will land on nodes with similar values for affinity labels.
```
same here

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,36 +633,60 @@ type ServiceAffinity struct {
 // 		- and SOME pod, from a service, in the same namespace, ALREADY scheduled onto a node, has a matching value.
-func (s *ServiceAffinity) CheckServiceAffinity(pod *api.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+
+// WARNING: This Predicate is NOT gauranteed to work if  some of the predicateMetadata data isn't precomputed...
```
s/if  some/if some/

wojtek-t commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -320,9 +320,14 @@ func (f *ConfigFactory) CreateFromKeys(predicateKeys, priorityKeys sets.String,
 	f.Run()
 
-	algo := scheduler.NewGenericScheduler(f.schedulerCache, predicateFuncs, priorityMetaProducer, priorityConfigs, extenders)
+	algo := scheduler.NewGenericScheduler(f.schedulerCache, predicateFuncs, priorityMetaProducer, predicateMetaProducer, priorityConfigs, extenders)
```
I would prefer to change the order: first predicate related, then priority-related. This reflects both: alphabetic order, as well as the order in which they are processed internally.

jayunit100 replied:
i think probably if thats the case ill reorg the prioconfigs to be one arg earlier to, so that the types are after the configs

jayunit100 commented:
@k8s-bot unit test this

jayunit100 commented:
ok, pushed some cleanup.  looks like i broke the scheduler_generic unit tests but fixed those to.

jayunit100 commented:
@k8s-bot gci gce e2e test this.

jayunit100 commented:
@k8s-bot cvm gce e2e test this

jayunit100 commented:
@k8s-bot gci gce e2e test this

jayunit100 commented:
@wojtek-t PTAL comments addressed... 
- you were right we can live without storing the extra metadata var,  i removed it.
- i didnt yet externalize the precompute function - are we sure its a good idea (i was thinking the purpose of New...Predicate is to provide the precomputation alongside the predicate, so good to keep them in the same code block.  but up to you.
- other minor issues addressed.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -0,0 +1,63 @@
+		podRequest:                     getResourceRequest(pod),
+		podPorts:                       getUsedPorts(pod),
+		matchingAntiAffinityTerms:      matchingTerms,
+		serviceAffinityMatchingPodList: nil,
```
If this is nil, the please remove it (it doesn't change anything).

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -0,0 +1,63 @@
+		matchingAntiAffinityTerms:      matchingTerms,
+		serviceAffinityMatchingPodList: nil,
+	}
+	if predicatePrecomputations != nil {
```
You don't need this if. In Go you can do "for ... := range x{" even if x is nil. This just works.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -36,6 +36,22 @@ import (
+
+var predicatePrecomputations map[string]PredicateMetadataModifier = make(map[string]PredicateMetadataModifier)
+
+var predicatePrecomputeRegisterLock sync.Mutex
```
nit: can you please move it above predicatePrecomputations? For me it's always easier to see the lock before the variable it's locking.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,20 +633,38 @@ type ServiceAffinity struct {
 // ServiceAffinity.labels are homogenous for pods added to a node.
-// (i.e. it returns true IFF this pod can be added to this node, such
-// that all other pods in the same service are running on nodes w/
+// (i.e. it returns true IFF this pod can be added to this nodesuch that all other pods in
```
s/nodesuch/node such/

wojtek-t commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -149,7 +150,14 @@ func findNodesThatFit(
 	predicateFuncs map[string]algorithm.FitPredicate,
-	extenders []algorithm.SchedulerExtender) ([]*api.Node, FailedPredicateMap, error) {
+	extenders []algorithm.SchedulerExtender,
+	metaFactory interface{},
```
```
metaProducer algorithm.MetadataProducer
```
[for naming consistency and avoid unnecessary casing].

wojtek-t commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -149,7 +150,14 @@ func findNodesThatFit(
+	extenders []algorithm.SchedulerExtender,
+	metaFactory interface{},
+) ([]*api.Node, FailedPredicateMap, error) {
+	// For the time being, the metadata factory is optional.  We may make it required at some point.
```
I think this is already kind-of required.

jayunit100 replied:
certainly, but at the function level i dont think we are throwing an error  if its nil ... henc the comment.  Should i throw an error if its nil ?

wojtek-t replied:
Aah - ok. Let's leave as is (i.e. not throw an error), just clarify the comment that it's about the returned value, not the factory itself.

wojtek-t commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -159,14 +167,15 @@ func findNodesThatFit(
 			nodeName := nodes[i].Name
-			fits, failedPredicates, err := podFitsOnNode(pod, meta, nodeNameToInfo[nodeName], predicateFuncs)
+			// pFactory takes a pod, node info, and creates a black-box type which can be used by the predicate.
+			fits, failedPredicates, err := podFitsOnNode(pod, pFactory(pod, nodeNameToInfo), nodeNameToInfo[nodeName], predicateFuncs)
```
Can can't call pFactory() for every node separately. You have to call it once at the beginining (where it used to be before).

jayunit100 replied:
"can can't" you mean - "can we?" :)  i think yes it would be an additional optimization

jayunit100 commented:
Finished rebasing and updating tests - some of the function names changed last night i guess and so it broke this PR.  Ready for another look :)

jayunit100 commented:
@k8s-bot gci gke e2e test this.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,20 +630,41 @@ type ServiceAffinity struct {
-func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister algorithm.ServiceLister, nodeInfo NodeInfo, labels []string) algorithm.FitPredicate {
+// serviceAffinityPrecomputation should be run once by the scheduler before looping through the Predicate.  It is a helper function that
+// only should be referenced by NewServiceAffinityPredicate.
+func serviceAffinityPrecomputation(serviceLister algorithm.ServiceLister, podLister algorithm.PodLister) func(pm *predicateMetadata) {
```
s/func(pm *predicateMetadata)/PredicateMetadataModifier/

jayunit100 replied:
agree, odd that this doesn't compile for me.` unexpected var, expecting expression` , let me see why

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,20 +630,41 @@ type ServiceAffinity struct {
+// serviceAffinityPrecomputation should be run once by the scheduler before looping through the Predicate.  It is a helper function that
+// only should be referenced by NewServiceAffinityPredicate.
+func serviceAffinityPrecomputation(serviceLister algorithm.ServiceLister, podLister algorithm.PodLister) func(pm *predicateMetadata) {
+	precomp := func(pm *predicateMetadata) {
```
Just:
```
return func(pm *predicateMetadata) {
...
```
Declaring a a variable for it is not necessary here.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,20 +630,41 @@ type ServiceAffinity struct {
+	precomp := func(pm *predicateMetadata) {
+		var err error
+		// Store services which match the pod.
+		pm.serviceAffinityMatchingPodServices, err = serviceLister.GetPodServices(pm.pod)
```
glog.Errorf if err != nil

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,20 +630,41 @@ type ServiceAffinity struct {
+		// consider only the pods that belong to the same namespace
+		allMatches, err := podLister.List(selector)
+		pm.serviceAffinityMatchingPodList = FilterPodsByNamespace(allMatches, pm.pod.Namespace)
+
```
nit: remove empty line

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -665,31 +698,20 @@ func (s *ServiceAffinity) CheckServiceAffinity(pod *api.Pod, meta interface{}, n
-			if len(nsServicePods) > 0 {
-				// consider any service pod and fetch the node its hosted on
-				otherNode, err := s.nodeInfo.GetNodeInfo(nsServicePods[0].Spec.NodeName)
+		if len(predicateMeta.serviceAffinityMatchingPodServices) > 0 {
```
Instead of lines 688-690 here what you should do here is to:
```
var services []*api.Service
if pm, ok := meta.(*predicateMetadata); ok {
  services = pm.serviceAffinityMatchingPodServices
} else {
  var err
  services, err = s.serviceLister.GetPodServices(pod)
  if err != nil {
    glog.Errorf(...)
  }
}
```

With this approach, this predicate will work no matter if the metadata is passed or not - the difference will be that with metadata is significantly more efficient.

wojtek-t commented on plugin/pkg/scheduler/algorithm/types.go:
```diff
@@ -54,10 +54,15 @@ type PriorityConfig struct {
 
-func EmptyMetadataProducer(pod *api.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo) interface{} {
-	return nil
+// emptyMetadataProducer returns a no-op MetadataProducer.
```
I don't understand why you did this change. It's more complcated now and not needed in my opinion.

jayunit100 replied:
i dug into it , i think its simpler if i just revert this change.   found a case yest where some casting was required that looked ugly to me, but its not going to effect the tests from passing

wojtek-t commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -153,7 +154,14 @@ func findNodesThatFit(
+) ([]*api.Node, FailedPredicateMap, error) {
+
+	pFactory := algorithm.EmptyMetadataProducer
+	if metaProducer == nil {
```
Why? Let's assume it is non-nil. This is pretty safe assumption.

wojtek-t commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -163,14 +171,16 @@ func findNodesThatFit(
 			nodeName := nodes[i].Name
-			fits, failedPredicates, err := podFitsOnNode(pod, meta, nodeNameToInfo[nodeName], predicateFuncs)
+			// pFactory takes a pod, node info, and creates a black-box type which can be used by the predicate.
+			metaDataProducer := pFactory(pod, nodeNameToInfo)
```
You are computing it for every node separate now. It doesn't make any sense. This should be done before loop.

jayunit100 replied:
this looks like an error on my end when i pushed it back up.  too bad i dont have a good unit test for this optimization.

jayunit100 commented:
@wojtek-t 
- Rather then adding the `GetPodServices` check inside `checkServiceAffinity`, Ive added a second commit which is separate, which adds a unit test that verifies predicate behaviour with / without precomputation.  That is a good stub for eventually making a decision on wether or not we want predicates to bootstrap their own data.    
- Other comments addressed - good catch on moving the pFactory() up, you are right that will speed things up also.  Regarding the EmptyMetadata change, i also reverted it - i thought it made casting logic less amibguous but probably adds no value.

jayunit100 commented on plugin/pkg/scheduler/algorithm/listers.go:
```diff
@@ -91,10 +91,6 @@ func (f FakeServiceLister) GetPodServices(pod *api.Pod) (services []*api.Service
 			services = append(services, service)
 		}
 	}
-	if len(services) == 0 {
```
This makes the if err == nil check which woj suggested less noisy,and more importantly makes it so semantics conform to the other GetPodsServices tests.

jayunit100 commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -650,46 +674,44 @@ func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister al
+func (s *ServiceAffinity) checkServiceAffinity(pod *api.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+	var services []*api.Service
+	var pods []*api.Pod
+	if pm, ok := meta.(*predicateMetadata); ok && pm.serviceAffinityMatchingPodList != nil && pm.serviceAffinityMatchingPodList != nil {
```
@wojtek-t 

You suggested making this bootstrap the services data... so  ive set it up so that at least the unit tests and this can easily be extended to support independently functional `checkServiceAffinity` at a later date.  I'd like to keep as-is for this PR though, b/c 

- Its somewhat noisy to do the query/check the error and is already done elsewhere, and it seems like a generally acceptable thing to expect predicate precomputations to be correct.
- Even if we did precompute serviecs, we also then have to check pods to make semantics consistent, which makes the logic of this method noisy/redundant code.
- If we want  want these functions to be able to query all of their data b/c some is missing, that is probably something we can do more elgantly by leveraging the  precomputations rather than on a case-by-case basis....

if you still feel strongly i can bootstrap the `GetPodServices` function in here, just my 2 cents for now...

jayunit100 commented on plugin/pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -1207,13 +1210,14 @@ func TestServiceAffinity(t *testing.T) {
+		labels         []string
+		fits           bool
+		test           string
+		skipPrecompute bool
```
As per comment above, IF this is specified as false, we can test that the function has the same results w/ or w/o precomputation.  That can be used later so that we can verify the function is 100% robust and independent of metadata, if we want to...

jayunit100 commented:
im giong to dig into the right semantics to keep for now wrt serices/  for now the unit test wont pass.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -650,46 +674,45 @@ func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister al
+		services = pm.serviceAffinityMatchingPodServices
+		pods = pm.serviceAffinityMatchingPodList
+	} else {
+		// TODO decide precomputation-data semantics, issue #33763.
```
It seems the issue number is wrong.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,20 +627,44 @@ type ServiceAffinity struct {
-func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister algorithm.ServiceLister, nodeInfo NodeInfo, labels []string) algorithm.FitPredicate {
+// serviceAffinityPrecomputation should be run once by the scheduler before looping through the Predicate.  It is a helper function that
+// only should be referenced by NewServiceAffinityPredicate.
+func serviceAffinityPrecomputation(serviceLister algorithm.ServiceLister, podLister algorithm.PodLister) PredicateMetadataModifier {
```
I really think that this should be a method on "ServiceAffinity" - in this case you won't need to pass any arguments to it.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -650,46 +674,45 @@ func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister al
+		pods = pm.serviceAffinityMatchingPodList
+	} else {
+		// TODO decide precomputation-data semantics, issue #33763.
+		return false, nil, fmt.Errorf("Precomputed data not available.")
```
Well, I think that what you should do here is:
```
tmp := predicateMetadata{pod: pod}
s.serviceAffinityPrecomputation(..)
serivces, pods = tmp.serviceAff..., tmp.serviceAff...
```

With this approach, the predicate work no matter what metadata you pass - it.s just more efficient when you pass metadata. And I don't see any problems with this approach.

jayunit100 replied:
Rerunning the whole predicate is also what i suggest in #35168, This is a superior approach to duplicating the exact logic.  for now i think thats ok, will impl.

wojtek-t reviewed (commented):
Few more (mostly minor) comments.

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,20 +627,44 @@ type ServiceAffinity struct {
+// serviceAffinityPrecomputation should be run once by the scheduler before looping through the Predicate.  It is a helper function that
+// only should be referenced by NewServiceAffinityPredicate.
+func (s *ServiceAffinity) serviceAffinityPrecomputation() PredicateMetadataModifier {
+	return func(pm *predicateMetadata) {
```
If you make ```predicateMetadata``` an argument of "serviceAffinityPrecomputation" function, then you don't need to return a function here - you can just use "serviceAffinityPrecomputation" below.

jayunit100 replied:
Yes but, (by definition) the precomputation takes a predicateMeta and mutates it...

wojtek-t replied:
I don't understand. You can simply have:
```
func (s *ServiceAffinity) serviceAffinityPrecomputation(pm *predicateMetadata) {
  // the body of this func
}
```

and this will work. It will be modifying the given pm.

wojtek-t commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -163,14 +166,15 @@ func findNodesThatFit(
 		var filteredLen int32
+
+		// We can use the same metadata producer for all nodes.
+		metaDataProducer := pFactory(pod, nodeNameToInfo)
```
s/metaDataProducer/meta/ (it is not a producer in fact).

wojtek-t commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -153,7 +154,9 @@ func findNodesThatFit(
 	predicateFuncs map[string]algorithm.FitPredicate,
-	extenders []algorithm.SchedulerExtender) ([]*api.Node, FailedPredicateMap, error) {
+	extenders []algorithm.SchedulerExtender,
+	pFactory algorithm.MetadataProducer,
```
nit: Can you call it metadataProducer instead pFactory (it would be more readable I think).

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,20 +627,45 @@ type ServiceAffinity struct {
-func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister algorithm.ServiceLister, nodeInfo NodeInfo, labels []string) algorithm.FitPredicate {
+// serviceAffinityPrecomputation should be run once by the scheduler before looping through the Predicate.  It is a helper function that
+// only should be referenced by NewServiceAffinityPredicate.
+
```
nit: remove empty line

wojtek-t commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -627,20 +627,45 @@ type ServiceAffinity struct {
 		labels:        labels,
 	}
-	return affinity.CheckServiceAffinity
+	return affinity.checkServiceAffinity, func(pm *predicateMetadata) {
```
just do:
```
return affinity.checkServiceAffinity, affinity.serviceAffinityPrecomputation
```

wojtek-t commented:
/lgtm

(I'm assuming that tests will pass)
